lib/SILOptimizer/Analysis/ArraySemantic.cpp

```diff
@@ -682,3 +684,36 @@ bool swift::ArraySemanticsCall::replaceByValue(SILValue V) {
   removeCall();
   return true;
 }
+
+bool swift::ArraySemanticsCall::replaceByAppendingValues(SILModule &M, SILFunction* appendFn, llvm::SmallVectorImpl<SILValue> &Vs, ArrayRef<ProtocolConformanceRef> &conformances) {
+  assert(getKind() == ArrayCallKind::kAppendContentsOf &&
+         "Must be an append_contentsOf call");
+  assert(appendFn && "Must provide an append SILFunction");
+
+  // We only handle loadable types.
+  for(auto& V : Vs) {
+    if (!V->getType().isLoadable(SemanticsCall->getModule()))
+      return false;
+  }
+
+  auto ArrRef = SemanticsCall->getArgument(1);
+  SILBuilderWithScope Builder(SemanticsCall);
+  auto *fnRef = Builder.createFunctionRef(SemanticsCall->getLoc(), appendFn);
+  auto fnTy = fnRef->getType();
+
+  for(auto& V : Vs) {
+    auto subTy = V->getType();
+    auto &ValLowering = Builder.getModule().getTypeLowering(subTy);
+    auto copiedVal = ValLowering.emitCopyValue(Builder, SemanticsCall->getLoc(), V);
+    auto allocStackInst = Builder.createAllocStack(SemanticsCall->getLoc(), subTy);
+    Builder.createStore(SemanticsCall->getLoc(), copiedVal, allocStackInst, StoreOwnershipQualifier::Unqualified);
+    ArrayRef<Substitution> subs{Substitution(subTy.getSwiftType(), conformances)};
```

Comment:
When I tried that, the return value was an empty `ArrayRef`. I figured that at this point, the original function has already been specialized, so the substitutions would be empty.